###### DEPURACION DE DATOS

#### 1-Importamos nuestras librerias

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns



#### 2- Importamos nuestros datos

In [2]:
# 2.1 Extraccion de datos 
df =  pd.read_csv('../data/attacks.csv', encoding = 'ISO-8859-1')

#### 3- Informacion sobre los datos

In [3]:
# 3.1.1 Comprobamos que se han importado con los primeros 5 registros 
df.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

In [4]:
# 3.1.2 Comprobamos que se han importado con los todos los registros.
df

Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
25718         NaN          NaN     NaN         NaN        NaN   
25719         NaN          NaN     NaN         NaN        NaN   
25720         NaN          NaN     NaN         NaN        NaN   
25721         NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
25718              NaN                             NaN          NaN   
25719              NaN                             NaN          NaN   
25720              NaN                             NaN          NaN   
25721              NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  Name Sex   ...         Species   \
0          Julie Wolfe    F  ...      White shark   
1      Adyson McNeely     F  ...              NaN   
2          John Denges    M  ...              NaN   
3                 male    M  ...        2 m shark   
4       Gustavo Ramos     M  ...  Tiger shark, 3m   
...                ...  ...  ...              ...   
25718              NaN  NaN  ...              NaN   
25719              NaN  NaN  ...              NaN   
25720              NaN  NaN  ...              NaN   
25721              NaN  NaN  ...              NaN   
25722              NaN  NaN  ...              NaN   

               Investigator or Source                       pdf  \
0                    R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1      K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2      K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3                      B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4                           A .Kipper      2018.06.04-Ramos.pdf   
...                               ...                       ...   
25718                             NaN                       NaN   
25719                             NaN                       NaN   
25720                             NaN                       NaN   
25721                             NaN                       NaN   
25722                             NaN                       NaN   

                                            href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...   
1      http://sharkattackfile.net/spreadsheets/pdf_di...   
2      http://sharkattackfile.net/spreadsheets/pdf_di...   
3      http://sharkattackfile.net/spreadsheets/pdf_di...   
4      http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                  ...   
25718                                                NaN   
25719                                                NaN   
25720                                                NaN   
25721                                                NaN   
25722                                                NaN   

                                                    href Case Number.1  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1      http://sharkattackfile.net/spreadsheets/pdf_di...    2

In [5]:
# 3.2 Informacion sobre numero de registro,variables de estudio(filas,columnas)
df.shape

(25723, 24)

#### OBSERVACION: Partimos de una muestra inicial de 25723 ataques registrados 



In [6]:
# 3.3 Obtencion de info sobre la naturaleza de las columnas 
df.info()  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

In [7]:
# 3.4.1 Obtencion de la suma de los valores nulos dentro de cada columna 
df.isnull().sum().sort_values(ascending=False)

Unnamed: 22               25722
Unnamed: 23               25721
Time                      22775
Species                   22259
Age                       22252
Sex                       19986
Activity                  19965
Location                  19961
Fatal (Y/N)               19960
Area                      19876
Name                      19631
Country                   19471
Injury                    19449
Investigator or Source    19438
Type                      19425
Year                      19423
href formula              19422
Date                      19421
pdf                       19421
href                      19421
Case Number.1             19421
Case Number.2             19421
original order            19414
Case Number               17021
dtype: int64

In [8]:
# 3.4.2 Obtencion del % de valores nulos dentro de cada columna y ordenamos de mayor a menor

ciento0 = df.isnull().sum().apply(lambda x: x/df.shape[0]).sort_values(ascending=False)

In [9]:
ciento0


Unnamed: 22               0.999961
Unnamed: 23               0.999922
Time                      0.885394
Species                   0.865335
Age                       0.865062
Sex                       0.776970
Activity                  0.776154
Location                  0.775998
Fatal (Y/N)               0.775959
Area                      0.772694
Name                      0.763169
Country                   0.756949
Injury                    0.756094
Investigator or Source    0.755666
Type                      0.755161
Year                      0.755083
href formula              0.755044
Date                      0.755005
pdf                       0.755005
href                      0.755005
Case Number.1             0.755005
Case Number.2             0.755005
original order            0.754733
Case Number               0.661704
dtype: float64

## analisis inicial 

### 4. Edicion de formato 

El principal objetivo de este punto es homojeinizar el formato de las columnas para poder acceder a ellas sin problemas 

In [10]:
# 4.1 analisis de columnas para ver "como estan escrito "
df.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

Se aprecia eterogeinidad en las etiquetas, series con espacios en blanco, mayusculas y minusculas entrmezcladas.
Esto podria iducir a errores al realizar consultas.

In [11]:
# 4.2 Edicion de nombres de columnas. Pasamaos la primera letra de todos los nombres a mayuscula
df.columns = df.columns.str.title()
# comprobacion 
df.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator Or Source', 'Pdf', 'Href Formula', 'Href',
       'Case Number.1', 'Case Number.2', 'Original Order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [12]:
# 4.3 Eleminamos los espacios en blanco 
df.columns = df.columns.str.strip()
#comprobacion 
df.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species', 'Investigator Or Source', 'Pdf', 'Href Formula', 'Href',
       'Case Number.1', 'Case Number.2', 'Original Order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [13]:
# 4.4 edicion de formato, remplazar caracteres
df.columns = df.columns.str.replace(" ", "_")
#comprobacion 
df.columns

Index(['Case_Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex', 'Age', 'Injury', 'Fatal_(Y/N)', 'Time',
       'Species', 'Investigator_Or_Source', 'Pdf', 'Href_Formula', 'Href',
       'Case_Number.1', 'Case_Number.2', 'Original_Order', 'Unnamed:_22',
       'Unnamed:_23'],
      dtype='object')

In [14]:
# 4.5 edicion de formato, remplazar y homogeinizar registros
df = df.apply(lambda x: x.str.strip() if x.dtype == "object" else x)
df = df.apply(lambda x: x.str.title() if x.dtype == "object" else x)
#comprobacion 
df["Fatal_(Y/N)"].value_counts()




N          4301
Y          1389
Unknown      71
M             1
2017          1
Name: Fatal_(Y/N), dtype: int64

#### 5. Eliminacion de Filas que CLARAMENTE no aporten valor 

In [15]:
# Antes de eliminar filas 
df.shape


(25723, 24)

In [16]:
# Al aplicar how ='all' aplicamos sobre todas las columnas con na.
df.dropna(how='all',inplace = True)

In [17]:
# Observamos como hemos quitado un 66% de registros sin valor alguno  
df.shape

(8703, 24)

#### 6. Eliminacion de columnas que sean CLARAMENTE innecesarias


In [18]:
# Usamos como guia el punto 3.4.2
ciento0

Unnamed: 22               0.999961
Unnamed: 23               0.999922
Time                      0.885394
Species                   0.865335
Age                       0.865062
Sex                       0.776970
Activity                  0.776154
Location                  0.775998
Fatal (Y/N)               0.775959
Area                      0.772694
Name                      0.763169
Country                   0.756949
Injury                    0.756094
Investigator or Source    0.755666
Type                      0.755161
Year                      0.755083
href formula              0.755044
Date                      0.755005
pdf                       0.755005
href                      0.755005
Case Number.1             0.755005
Case Number.2             0.755005
original order            0.754733
Case Number               0.661704
dtype: float64

In [22]:
# 6.1 Usamos .drop para eliminar las columnas que nos den igual y con inplace = False lo hacemos permanentemente
df.drop(['Investigator_Or_Source', 'Pdf', 'Href_Formula', 'Href',
       'Case_Number.1', 'Case_Number.2', 'Original_Order', 'Unnamed:_22',
       'Unnamed:_23'], axis=1, inplace = True)

In [23]:
#comprobacion de 6.1
df.columns

Index(['Case_Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex', 'Age', 'Injury', 'Fatal_(Y/N)', 'Time',
       'Species'],
      dtype='object')

In [24]:
#comprobacion de 6.1
df.shape

(8703, 15)

In [25]:
df

Case_Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        Usa   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        Usa   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        Usa   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  Australia   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     Mexico   
...           ...          ...     ...         ...        ...   
8698            0          NaN     NaN         NaN        NaN   
8699            0          NaN     NaN         NaN        NaN   
8700            0          NaN     NaN         NaN        NaN   
8701            0          NaN     NaN         NaN        NaN   
25722          Xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free Diving   
...                ...                             ...          ...   
8698               NaN                             NaN          NaN   
8699               NaN                             NaN          NaN   
8700               NaN                             NaN          NaN   
8701               NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                 Name  Sex  Age  \
0         Julie Wolfe    F   57   
1      Adyson Mcneely    F   11   
2         John Denges    M   48   
3                Male    M  NaN   
4       Gustavo Ramos    M  NaN   
...               ...  ...  ...   
8698              NaN  NaN  NaN   
8699              NaN  NaN  NaN   
8700              NaN  NaN  NaN   
8701              NaN  NaN  NaN   
25722             NaN  NaN  NaN   

                                                  Injury Fatal_(Y/N)  \
0      No Injury To Occupant, Outrigger Canoe And Pad...           N   
1                             Minor Injury To Left Thigh           N   
2           Injury To Left Lower Leg From Surfboard Skeg           N   
3                              Minor Injury To Lower Leg           N   
4      Lacerations To Leg & Hand Shark Provoked Incident           N   
...                                                  ...         ...   
8698                                                 NaN         NaN   
8699                                                 NaN         NaN   
8700                                                 NaN         NaN   
8701                                                 NaN         NaN   
25722                                                NaN         NaN   

                Time          Species  
0              18H00      White Shark  
1      14H00  -15H00              NaN  
2              07H45              NaN  
3                NaN        2 M Shark  
4                NaN  Tiger Shark, 3M  
...              ...              ...  
8698             NaN              NaN  
8699             NaN              NaN  
8700             NaN              NaN  
8701             NaN              NaN  
25722            NaN              NaN  

[8703 rows x 15 columns]

nota: parece que hay mucho caso 0 Case Number. Procedemos a comprobar casos duplicados en Case Number.

##### 7.Analisis duplicados 

In [27]:
# usamos el metodod .duplicated() y luego sumamos su aparicion en la serie case number.
df['Case_Number'].duplicated().sum() 

2415

In [28]:
# usamos el metodo df.drop_duplicates() y reasignamos a df.
df = df.drop_duplicates()

In [29]:
# comprobamos que ya no esten.
df.duplicated().sum()

0

In [30]:
df

Case_Number         Date    Year        Type             Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating                 Usa   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked                 Usa   
2      2018.06.09  09-Jun-2018  2018.0     Invalid                 Usa   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked           Australia   
4      2018.06.04  04-Jun-2018  2018.0    Provoked              Mexico   
...           ...          ...     ...         ...                 ...   
6299      Nd.0003    1900-1905     0.0  Unprovoked                 Usa   
6300      Nd.0002    1883-1889     0.0  Unprovoked              Panama   
6301      Nd.0001    1845-1853     0.0  Unprovoked  Ceylon (Sri Lanka)   
6302            0          NaN     NaN         NaN                 NaN   
25722          Xx          NaN     NaN         NaN                 NaN   

                   Area                             Location     Activity  \
0            California          Oceanside, San Diego County     Paddling   
1               Georgia       St. Simon Island, Glynn County     Standing   
2                Hawaii                         Habush, Oahu      Surfing   
3       New South Wales                   Arrawarra Headland      Surfing   
4                Colima                             La Ticla  Free Diving   
...                 ...                                  ...          ...   
6299     North Carolina                       Ocracoke Inlet     Swimming   
6300                NaN                 Panama Bay 8ºn, 79ºw          NaN   
6301   Eastern Province  Below The English Fort, Trincomalee     Swimming   
6302                NaN                                  NaN          NaN   
25722               NaN                                  NaN          NaN   

                        Name  Sex  Age  \
0                Julie Wolfe    F   57   
1             Adyson Mcneely    F   11   
2                John Denges    M   48   
3                       Male    M  NaN   
4              Gustavo Ramos    M  NaN   
...                      ...  ...  ...   
6299   Coast Guard Personnel    M  NaN   
6300         Jules Patterson    M  NaN   
6301                    Male    M   15   
6302                     NaN  NaN  NaN   
25722                    NaN  NaN  NaN   

                                                  Injury Fatal_(Y/N)  \
0      No Injury To Occupant, Outrigger Canoe And Pad...           N   
1                             Minor Injury To Left Thigh           N   
2           Injury To Left Lower Leg From Surfboard Skeg           N   
3                              Minor Injury To Lower Leg           N   
4      Lacerations To Leg & Hand Shark Provoked Incident           N   
...                                                  ...         ...   
6299                                               Fatal           Y   
6300                                               Fatal           Y   
6301   Fatal. "Shark Bit Him In Half, Carrying Away T...           Y   
6302                                                 NaN         NaN   
25722                                                NaN         NaN   

                Time          Species  
0              18H00      White Shark  
1      14H00  -15H00              NaN  
2              07H45              NaN  
3                NaN        2 M Shark  
4                NaN  Tiger Shark, 3M  
...              ...              ...  
6299             NaN              NaN  
6300             NaN              NaN  
6301             NaN              NaN  
6302             NaN              NaN  
25722            NaN              NaN  

[6304 rows x 15 columns]

#### 8. Exportamos nuevo Df

In [31]:
df.to_csv("../data/sharks-clean.csv", encoding="latin1")